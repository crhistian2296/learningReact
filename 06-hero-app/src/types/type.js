const type = {
  login: '[auth]: Login',
  logout: '[auth]: Logout',
};

export const { login, logout } = type;
