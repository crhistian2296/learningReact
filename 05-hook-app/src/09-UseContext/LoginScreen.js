import { useContext } from 'react';
import { UserContext } from './UserContext';

const LoginScreen = () => {
  const { user, setUser } = useContext(UserContext);
  // console.log(user);
  return (
    <div>
      <h1>LoginScreen</h1>
      <hr />
      <button
        className='btn btn-primary'
        onClick={() => {
          setUser({ id: 1212, name: 'Paco' });
        }}
      >
        Login
      </button>
      <button
        className='btn btn-primary mx-2'
        onClick={() => {
          setUser({});
        }}
      >
        Logut
      </button>
      <pre>{JSON.stringify(user, null, 3)}</pre>
    </div>
  );
};

export default LoginScreen;
