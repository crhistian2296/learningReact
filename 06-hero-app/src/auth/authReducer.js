// const user = {
//   name:'Fernando',
//   logged: true
// }

import { login, logout } from '../types/type';

export const authReducer = (state = {}, action) => {
  switch (action.type) {
    case login:
      return {
        ...action.payload,
        logged: true,
      };
    case logout:
      return {
        logged: false,
      };
    default:
      return state;
  }
};
