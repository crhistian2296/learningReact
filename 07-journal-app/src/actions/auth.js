import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types';
import { uiFinishLoading, uiStartLoading } from './ui';

// export const startLoginEmailPassword = (email, password) => {
//   // *redux-thunk provee automaticamente ese dispatch para acciones asincronas
//   return (dispatch) => {
//     setTimeout(() => {
//       dispatch(login(123, 'crhistian'));
//     }, 3000);
//   };
// };

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});
/**
 * Crea una cuenta en la app de autenticacion en firebase mediante correo
 * @param {string} email
 * @param {string} password
 * @param {string} name
 * @returns void
 */
export const startRegisterWithEmailAsync = (email, password, name) => {
  return (dispatch) => {
    dispatch(uiStartLoading());
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        console.log(user);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => console.log(err))
      .finally(() => dispatch(uiFinishLoading()));
  };
};
export const startLoginWithEmailAsync = (email, password) => {
  return (dispatch) => {
    dispatch(uiStartLoading());
    firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        console.log(user);
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => console.log(err))
      .finally(() => dispatch(uiFinishLoading()));
  };
};

/**
 * Crean una cuenta en la app de autenticacion de firebase mediante boton de google
 * @returns void
 */
export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(googleAuthProvider)
      .then(({ user }) => dispatch(login(user.uid, user.displayName)))
      .catch(console.log);
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();

    await dispatch(logout());
  };
};

export const logout = () => ({ type: types.logout });
