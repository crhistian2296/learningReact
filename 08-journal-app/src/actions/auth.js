import { login } from '../types/types';

export const loginAction = (uid, displayName) => ({
  type: login,
  payload: {
    uid,
    displayName
  }
})
