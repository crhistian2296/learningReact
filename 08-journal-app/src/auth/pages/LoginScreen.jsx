import { Google } from '@mui/icons-material';
import {
  Alert,
  Button,
  Grid,
  Link,
  TextField,
  Typography,
} from '@mui/material';
import { useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link as RouteLink } from 'react-router-dom';
import { useForm } from '../../hooks';
import {
  startGoogleSignIn,
  startLoginWithEmailAndPassword,
} from '../../store/auth';
import AuthLayout from '../layout/AuthLayout';

const formData = {
  email: 'user@gmail.com',
  password: '123456',
};

const Login = () => {
  const { email, password, onInputChange } = useForm(formData);

  const { status, errorMessage = null } = useSelector(state => state.auth);
  const dispatch = useDispatch();

  const isAuthenticating = useMemo(() => status === 'checking', [status]);

  const onSubmit = event => {
    event.preventDefault();
    if (!email || !password) return console.warn('Incomplete credentials');

    console.debug('Credentials', { email, password });
    return dispatch(startLoginWithEmailAndPassword({ email, password }));
  };

  const onGoogleSignIn = async () => {
    console.debug('Google sign in');
    return dispatch(startGoogleSignIn());
  };

  return (
    <AuthLayout title='Login'>
      <form
        aria-label='submit-form'
        onSubmit={onSubmit}
        className='animate__animated animate__fadeIn animate__faster'
      >
        <Grid container>
          <Grid item xs={12} mt={2}>
            <TextField
              label='mail'
              type='email'
              fullWidth
              autoComplete='email'
              name='email'
              value={email}
              onChange={onInputChange}
            />
          </Grid>
          <Grid item xs={12} mt={2}>
            <TextField
              label='password'
              type='password'
              fullWidth
              autoComplete='current-password'
              name='password'
              inputProps={{ 'data-testid': 'password' }}
              value={password}
              onChange={onInputChange}
            />
          </Grid>
        </Grid>
        <Grid container spacing={2} mt={1}>
          <Grid
            item
            xs={12}
            sm={12}
            display={errorMessage !== null ? '' : 'none'}
          >
            <Alert severity='error'>{errorMessage}</Alert>
          </Grid>
          <Grid item xs={12} sm={6}>
            <Button
              type='submit'
              disabled={isAuthenticating}
              variant='contained'
              fullWidth
            >
              Login
            </Button>
          </Grid>
          <Grid item xs={12} sm={6}>
            <Button
              disabled={isAuthenticating}
              onClick={onGoogleSignIn}
              aria-label='google-btn'
              variant='contained'
              fullWidth
            >
              <Google />
              <Typography marginLeft={1} variant='body2'>
                Google
              </Typography>
            </Button>
          </Grid>
          <Grid container direction='row' justifyContent='end'>
            <Link
              component={RouteLink}
              color='inherit'
              mt={1}
              to='/auth/register'
            >
              Create account
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  );
};

export default Login;
