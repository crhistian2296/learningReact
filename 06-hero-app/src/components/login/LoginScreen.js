import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';
import { login } from '../../types/type';

const LoginScreen = () => {
  // react-router hook
  const navigate = useNavigate();
  const { dispatch } = useContext(AuthContext);

  const loginAction = {
    type: login,
    payload: {
      name: 'crhistian2296',
    },
  };

  const handleLogin = () => {
    dispatch(loginAction);

    const lastPath = localStorage.getItem('lastPath') ?? '/';

    navigate(lastPath, { replace: true });
  };

  return (
    <div className='m-5'>
      <h1>LoginScreen</h1>
      <hr />
      <button className='btn btn-primary' onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};

export default LoginScreen;
