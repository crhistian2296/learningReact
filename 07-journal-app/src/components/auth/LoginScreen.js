import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import useForm from '../../hooks/useFrom';

const LoginScreen = () => {
  const { formValues, handleInputChange } = useForm({
    email: 'cristof@gmail.com',
    password: '1234',
  });

  const { email, password } = formValues;

  const dispatch = useDispatch();

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(123456, 'cristof'));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h3 className='box-container__title'>Login</h3>

      <form onSubmit={handleLogin}>
        <input
          type='text'
          placeholder='Email'
          name='email'
          className='box-container__input'
          value={email}
          onChange={handleInputChange}
          autoComplete='off'
        />
        <input
          type='password'
          placeholder='Password'
          name='password'
          className='box-container__input'
          value={password}
          onChange={handleInputChange}
        />
        <button type='submit' className='btn btn-primary btn-block mt-1'>
          Login
        </button>
        <div className='box-container__social-networks'>
          <p>Login with social networks</p>

          <div className='google-btn' onClick={handleGoogleLogin}>
            <div className='google-icon-wrapper'>
              <img
                className='google-icon'
                src='https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg'
                alt='google button'
              />
            </div>
            <p className='google-btn-text'>
              <b>Sign in with google</b>
            </p>
          </div>
        </div>
        <Link to='/auth/register' className='link'>
          Create new account
        </Link>
      </form>
    </>
  );
};

export default LoginScreen;
