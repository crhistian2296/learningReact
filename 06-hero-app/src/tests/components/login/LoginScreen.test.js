import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../../auth/AuthContext';
import LoginScreen from '../../../components/login/LoginScreen';
import { login } from '../../../types/type';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas en <LoginScreen/>', () => {
  const contextValue = {
    user: {
      logged: false,
    },
    dispatch: jest.fn(),
  };

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter initialEntries={['/login']}>
        <LoginScreen />
        {/* <Routes>
          <Route path='/login' element={<LoginScreen />}></Route>
        </Routes> */}
      </MemoryRouter>
    </AuthContext.Provider>
  );

  test('Debe de hacer Match cons la snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Debe de realizar el dispatch y la navegacion', () => {
    const { dispatch } = contextValue;

    const loginAction = {
      type: login,
      payload: {
        name: 'crhistian2296',
      },
    };

    wrapper.find('button').simulate('click');

    expect(dispatch).toHaveBeenCalledWith(loginAction);
    expect(mockNavigate).toHaveBeenCalledWith('/', { replace: true });

    localStorage.setItem('lastPath', '/dc');
    wrapper.find('button').simulate('click');

    expect(mockNavigate).toHaveBeenCalledWith('/dc', { replace: true });
  });
});
