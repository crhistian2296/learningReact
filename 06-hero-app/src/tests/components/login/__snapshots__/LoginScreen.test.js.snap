// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen/> Debe de hacer Match cons la snapshot 1`] = `
<div
  className="m-5"
>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
